crates/oxc_semantic/src/scope.rs

```diff
@@ -184,16 +184,20 @@ impl ScopeTree {
         let scope_id = self.parent_ids.push(parent_id);
         _ = self.flags.push(flags);
         _ = self.bindings.push(Bindings::default());
+        // XXX: should this be a real invalid sentinel value or an Option?
+        _ = self.node_ids.push(AstNodeId::dummy());
```

Comment:
I think sentinel (as you have it now) is better. `Option<AstNodeId>` would be 16 bytes, rather than 8.

Once #4203 is merged, node IDs will be generated before scope is created, so we can add `node_id` to params of this function, and push the actual `AstNodeId` here. Then we can remove `set_node_id`.

Reply:
Ok, I've dropped the comment.

My concerns were that:
 * `dummy()` isn't a real **invalid** sentinel value. To the contrary, ID 0 is the most common valid ID. This setup seems prone to hard-to-debug logic bugs.
 * `AstNodeId` is not well-suited for an `Option<AstNodeId>` as it does not support niche-optimization (as you pointed out).

Maybe longer term this is going to be a moot point, but I guess it makes sense to turn `AstNodeId` into a `NonZeroUsize`.

Reply:
Yes, I agree with your concerns. It's a bit messy.

The background which I imagine you're not aware of because the info was buried in various issues is that I'm planning to do a "holistic" review of `Semantic` and make an RFC to significantly revamp it. This kind of issue is one I'll be looking at. That review will take me a month or so - first I need to familiarize myself with all the ways `Semantic` is used in both Oxc and Rolldown (there's a lot of Oxc I don't know at present - I've never touched the linter, for example).

But the perf improvements you've been working on were definitely going to be part of my conclusions, so getting them done now, in advance of my RFC, is really helpful as (a) it gives us perf boost *now*, and (b) it narrows the scope of what the RFC will have to cover.